### PR TITLE
[SPARK-28250][SQL] QueryPlan#references should exclude producedAttributes

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a followup of the discussion in https://github.com/apache/spark/pull/24675#discussion_r286786053

`QueryPlan#references` is an important property. The `ColumnPrunning` rule relies on it.

Some query plan nodes have `Seq[Attribute]` parameter, which is used as its output attributes. For example, leaf nodes, `Generate`, `MapPartitionsInPandas`, etc. These nodes override `producedAttributes` to make `missingInputs` correct.

However, these nodes also need to override `references` to make column pruning work. This PR proposes to exclude `producedAttributes` from the default implementation of `QueryPlan#references`, so that we don't need to override `references` in all these nodes.

Note that, technically we can remove `producedAttributes` and always ask query plan nodes to override `references`. But I do find the code can be simpler with `producedAttributes` in some places, where there is a base class for some specific query plan nodes.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gatorsmile @viirya @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 1]
```diff
@@ -59,12 +51,21 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
...
+   * include attributes that are implicitly referenced by being passed through to the output tuple.
+   */
+  @transient
+  lazy val references: AttributeSet = {
+    AttributeSet.fromAttributeSets(expressions.map(_.references)) -- producedAttributes
+  }
+
   /**
    * Attributes that are referenced by expressions but not provided by this node's children.
    * Subclasses should override this method if they produce attributes internally as it is used by
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: We can't override this now.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: It seems like cheating if we override this method. We want to know if the plan is valid or not by this method, overriding this method means we skip this validation.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: Ok. I think it makes sense to me. We should update this comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 2]
```diff
@@ -451,8 +451,6 @@ case class FlatMapGroupsInRExec(
 
   override def outputPartitioning: Partitioning = child.outputPartitioning
 
-  override def producedAttributes: AttributeSet = AttributeSet(outputObjAttr)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: Isn't `outputObjAttr` produced by `FlatMapGroupsInRExec`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: This is a no-op override, it's the same as `ObjectProducerExec.producedAttributes`

**Review by viirya — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**